### PR TITLE
Improve the stability of WebDriver when Selenium server is gone

### PR DESCRIPTION
**Pull request by eXorus**
Fix issue #3531 

**Step to reproduce :**
- Launch Selenium Server
- Launch Codeception (Acceptance Test)
- Stop Selenium Server

Each time I stopped server  during the $I->wait() and before $I->acceptPopup()
This is all the the fix I need to do to be able to have an Codeception that never exit and that finish all the tests and give the feedback about how many errors ....

Can solved this issue:

- Failed to connect to localhost port 4446: Connection refused
- Error forwarding the request 172.17.0.4:5555 failed to respond
- Session xxx was terminated due to FORWARDING_TO_NODE_FAILED

	------------------------
	Exception Found
	WebDriver.php:xxx
	WebDriver.php:xxx

	Fix x
	method that throw exception => description what it does before fix
	description what it does after fix

	------------------------
	WebDriverCurlException
	WebDriver.php:389
	WebDriver.php:374

	Fix 1
	debugWebDriverLogs => catch only UnknownServerException for IE
	so now catch all exception

	------------------------
	WebDriverCurlException
	WebDriver.php:513
	WebDriver.php:377

	Fix 2
	_saveScreenshot => no catch, bad design
	so now I'm using Exception to trap it

	------------------------
	WebDriverCurlException
	WebDriver.php:537
	WebDriver.php:378

	Fix 3
	_savePageSource => no catch, bad design
	so now I'm using Exception to trap it

	------------------------
	WebDriverCurlException
	WebDriver.php:439
	WebDriver.php:364

	Fix 4
	cleanWebDriver => catch only UnknownServerException
	so now I'm using Exception to trap it

	------------------------
	ConnectionException
	WebDriver.php:308
	WebDriver.php:324

	Fix 5
	_initialize => catch only WebDriverCurlException to throw ConnectionException
	so now I'm replacing to catch all Exception, move the catch at the end and remove new exception

The last fix is not enough good for me:
If I throw ConnectionException, the script exit => bad
If I only print debug line => the test is runned and failed because WebDriver is null ([PHPUnit_Framework_Exception] Undefined property: Codeception\Module\WebDriver::$webDriver)

The best way was to mark the test has skipped but I didn't succeed to call this function in WebDriver module.

### CODE REVIEW

File: src/Codeception/Module/WebDriver.php [hunk 1]
```diff
@@ -304,13 +304,11 @@ public function _initialize()
                 $this->httpProxyPort
             );
             $this->sessions[] = $this->_backupSession();
-        } catch (WebDriverCurlException $e) {
-            throw new ConnectionException(
-                $e->getMessage() . "\n \nPlease make sure that Selenium Server or PhantomJS is running."
-            );
+            $this->webDriver->manage()->timeouts()->implicitlyWait($this->config['wait']);
+            $this->initialWindowSize();
```
Comment (Naktibalda) [inline src/Codeception/Module/WebDriver.php]: Could you explain these 2 lines?

File: src/Codeception/Module/WebDriver.php [hunk 2]
```diff
@@ -511,11 +507,13 @@ public function _getCurrentUri()
 
     public function _saveScreenshot($filename)
     {
-        if ($this->webDriver !== null) {
+        try {
+            if ($this->webDriver === null) {
+                throw new \Exception("WebDriver::_saveScreenshot method has been called when webDriver is not set");
```
Comment (Naktibalda) [inline src/Codeception/Module/WebDriver.php]: I am not sure if it is a very good idea to throw exception and catch it immediately,
wouldn't it be better to print some debug output and return?

**Comment by Naktibalda** [comment]
@eXorus I have added spaces around concatenation operator, please pull before making more changes.

File: src/Codeception/Module/WebDriver.php (hunk 1 above)
Comment (eXorus) [inline src/Codeception/Module/WebDriver.php]: These 2 lines was already existing but I put them in the try to catch exception from Selenium.

File: src/Codeception/Module/WebDriver.php (hunk 2 above)
Comment (eXorus) [inline src/Codeception/Module/WebDriver.php]: I agree with you but I didn't know how to do that. I don't want to repeat the error.

```php
try {
if ($this->webDriver === null) {
throw new \Exception("WebDriver::_saveScreenshot method has been called when webDriver is not set");
}
$this->webDriver->takeScreenshot($filename);
 } catch (\Exception $e) {
 $this->debug("Unable to retrieve screenshot from Selenium : ".$e->getMessage());
}
```

or I can do that if you prefer
```php
if ($this->webDriver === null) {
$this->debug("WebDriver::_saveScreenshot method has been called when webDriver is not set");
}

try {
$this->webDriver->takeScreenshot($filename);
 } catch (\Exception $e) {
 $this->debug("Unable to retrieve screenshot from Selenium : ".$e->getMessage());
}
```

**Comment by eXorus** [comment]
And could you help me for the skipped test in initialize() ? because for the moment this PR can't be merged.

File: src/Codeception/Module/WebDriver.php (hunk 2 above)
Comment (Naktibalda) [inline src/Codeception/Module/WebDriver.php]:
```php
if ($this->webDriver === null) {
    $this->debug("WebDriver::_saveScreenshot method has been called when webDriver is not set");
    return;
}
```

File: src/Codeception/Module/WebDriver.php (hunk 1 above)
Comment (Naktibalda) [inline src/Codeception/Module/WebDriver.php]: oh, it is fine then.

**Comment by Naktibalda** [comment]
I updated `_before` method, please test.

**Comment by eXorus** [comment]
It's working well:
The first one failed: [Facebook\WebDriver\Exception\UnknownServerException] cannot forward the request Connect to 172.17.0.6:5555 [/172.17.0.6] failed: Connection refused
And the others are skipped.
Tests: 101, Assertions: 597, Errors: 2, Skipped: 91.

Tests are OK for me

**Comment by Naktibalda** [comment]
I think that it would be better to allow _initialize fail hard. Because it usually means that Selenium is not running.

This PR must be approved by @davert.

**Comment by eXorus** [comment]
I understand it's a choice to do. 
Fail hard or fail for the first and skip the others tests.

**Comment by DavertMik** [comment]
Thanks. 
I made a few tries and I think this can't be done in current way. 
If selenium is not running - all tests are skipped, no errors. This means that the overall result is OK, and no failures in exit code. It's good to skip other tests but at least one test should fail to indicate problem

**Comment by eXorus** [comment]
Sorry I'm not able to do it.
I don't know how to failed $test in the method _before() in the module WebDriver.

Any help ?

**Comment by Naktibalda** [comment]
I think that we have to introduce new methods `setFail` and `getFail` to 
[Codeception\Test\Metadata](https://github.com/Codeception/Codeception/blob/0e268a10e2e5c7d640a70ddf7dc7d558f2d3b0c1/src/Codeception/Test/Metadata.php), similar to `setSkip` and `setIncomplete`.

A handling of failure should be added to [IgnoreIfMetadataBlocked feature](https://github.com/Codeception/Codeception/blob/0e268a10e2e5c7d640a70ddf7dc7d558f2d3b0c1/src/Codeception/Test/Feature/IgnoreIfMetadataBlocked.php#L20) and [Unit format](https://github.com/Codeception/Codeception/blob/146b58e6d53008dbe289ab59ae9dd7cb80415c65/src/Codeception/Test/Unit.php#L34).

**Comment by eXorus** [comment]
lol, I just begin to do that before you read ... I will continue

**Comment by Naktibalda** [comment]
@DavertMik it is good now, right?

**Comment by DavertMik** [comment]
There are some internal changes. I see what you are up to but probably I have a better solution. https://github.com/Codeception/Codeception/pull/3598 This will allow us to throw exceptions in _before if it is needed. 

Please revert your changes to metadata behavior and merge with my branch (or just copy-paste my code, actually). Then WebDriver._before will be able just to throw exception in case a connection can't be established. `$this->fail` would be ok

**Comment by DavertMik** [comment]
Also it should be noted and not forgotten. This breaks AngularJS module. Execute `codecept run angular` to see output

**Comment by DavertMik** [comment]
Thanks! That was a long story but finally it is ready to be merged :+1:

**Comment by eXorus** [comment]
Thanks for your help @DavertMik and @Naktibalda 
Each day I learn more about Codeception :)

**Pull request merged**